app/models/user.rb

```diff
@@ -1,2 +1,36 @@
+require 'sysrandom'
+require 'openssl'
+# TODO: implement way of using external public key for encryption.
+EXAMPLE_RSA_KEY = '-----BEGIN PUBLIC KEY-----
+MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCXmtTTvovu9keHY//4RFoCuBmh
+1UdEnDGsOef3543eABZz4lqr5QNO0ktfRg5kwheqVwZJHshYZuZzfgFOxVD8YlG4
+F7kvhkNaC2M0XGM2BuQBgqrkOMB0EmhqBROP6tAQ4DLfTAm8FAo/rryrRoAejwxq
+A8ZHdCZuINXTJluGcQIDAQAB
+-----END PUBLIC KEY-----
+
+-----BEGIN RSA PRIVATE KEY-----
+MIICXAIBAAKBgQCXmtTTvovu9keHY//4RFoCuBmh1UdEnDGsOef3543eABZz4lqr
+5QNO0ktfRg5kwheqVwZJHshYZuZzfgFOxVD8YlG4F7kvhkNaC2M0XGM2BuQBgqrk
+OMB0EmhqBROP6tAQ4DLfTAm8FAo/rryrRoAejwxqA8ZHdCZuINXTJluGcQIDAQAB
+AoGADGbE/CE9LyoEReX/PV7SAiDvL9bTgqBIS+9cs2HEsCFPtOESS7drItp/Aeco
+LEPYTb3arMaQ2vJn7ZCkKL5pVmuzEvFXjO3vY0283U1xUdoevL/2yyHAM3l2dYbp
+bvsPrWZhoa06mqcmN1d4XPKnLLrJzYOWDvlW7IXop6Y4EvECQQDJwHpY+bao5lE+
+S7I3FnAzu6jUYA2VCTWDJVilX8sc/3OhONFlfAuv4gC2Yh+M4n8JhDZaKJ5q4Dbd
+50jzSPJ9AkEAwF5/YneLUldjYZfpj/8dDBkSZ1qwB5kBpM0q/8UPassBVP00A3pa
+4PCvxqhcsDPhnM9KX8N0D1f8fIM2vAoSBQJADW5l9sRxF591qysI60fwIlGM9M6M
+OO7LBDcZRrEw6XZMMWCB+Dqg4I2wNHQl25A8mtqVb70ihsT0giNipOenVQJBALyp
+cOEPQPJGhk8mf4gbuU8uH6CBjnZFFxmbpfomN9KSugGyOhrXGt3U728zm4Orvi7e
+FaEJvX922Ub8IOgWMEkCQDAUHoY4SEZQPI9PpQrcJP87J4x3+8CfWni6SAWkuKUU
+BsDoI6qtrBbVvkxx/wzv/i28scwKCj4enQkFbDWjPxg=
+-----END RSA PRIVATE KEY-----'.freeze
+
 class User < ApplicationRecord
+  before_save :generate_password
+  validates :username, uniqueness: true
+  def generate_password
+    clear_pass = password.nil? || password.empty? ? Sysrandom.base64(9) : password
+    key = OpenSSL::PKey::RSA.new EXAMPLE_RSA_KEY
```

Comment:
It works, but no. My task definition was maybe not very clear and I'm sorry for that. I'm looking for a simpler encryption solution that uses a predefined `secret` for encryption and decryption. 

Something more like this https://gist.github.com/0sc/2e2bdb19bcc0051e715625075e7cc068. But even simpler. In the gist the key is generated from random salts whereas we need to use a fixed key which is passed to the Rails environment through e.g. the `SECRET_KEY` environment variable.

This secret shall in turn be used in KADOS to decrypt the base64 encoded and encrypted password stored in the database.

The secret has to be exactly 32 char long for AES encryption but I've usually cheated and generated a sha256 from the secret to obtain a 32 char long secret. Hence the user can decide to use a short and bas password like `pass123` or a long random password like `NmQxYmIwN2Y1NjgxNGRmYWYzN2FkOTY4YTJiY2VkMjRmMDFlNmNjNGZlY2M2ZTE4` and I always get the fixed 32 char secret.

I recently picked AES-256-GCM as a cipher in another project I created and if you have time you could explore if this alternative works in both Ruby and PHP.

Reply:
I experimented with AES-256-GCM yesterday and found that Ruby, PHP and Golang all seems to implement encryption/decryption differently. So far I've only been able to encrypt and decrypt a secret within one language. Trying to encrypt in e.g. Ruby and decrypt in PHP is still a mystery how to get it working.

To be continued ...

Reply:
So now I managed to encrypt with Ruby and decrypt with PHP.

```ruby

Reply:
The catch was to find the right option in PHP. I figured OPENSSL_RAW_DATA was the default so  tested a bunch of other flags and combination. Then I decided to try OPENSSL_RAW_DATA and got it working 🎉

Reply:
All right, I'll switch the encryption to AES-GCM. I'll have to look around a bit how to implement secrets in rails and I'll do that aswell.

Reply:
With docker and kubernetes secrets and other configuration parameters are given as environment variables, thus you just have to use `ENV["SOME_VARIABLE_NAME"]` in your code.